backend/src/controllers/upload.ts

```diff
@@ -1,117 +1,264 @@
-import express from "express";
+import { Upload } from "@contract";
 import { Storage } from "@google-cloud/storage";
-import stream from "stream";
+import crypto from "crypto";
+import express from "express";
+import FileType from "file-type";
+import mime from "mime-types";
+import {
+  BadRequestError,
+  Get,
+  JsonController,
+  NotFoundError,
+  Post,
+  Req,
+  UseBefore,
+} from "routing-controllers";
+import sharp from "sharp";
 import { Readable } from "stream";
+import { z } from "zod";
 import fullConfig from "../config/config";
-import crypto from "crypto";
-import { Upload } from "../models/upload";
+import { IsLoggedInMiddleware } from "../helpers/auth";
+import { Upload as UploadModel } from "../models/upload";
+import { zodBase64FileSchema } from "../utils/validation";
+
+const MAX_FILE_SIZE_MB = 10;
+const MAX_IMAGE_SIZE_MB = 5;
+const ALLOWED_FILE_MIME_TYPES = [
+  "image/png",
+  "image/jpeg",
+  "video/mp4",
+  "application/pdf",
+  "application/msword",
+  "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
+  "text/plain",
+];
+const ALLOWED_IMAGE_MIME_TYPES = ["image/png", "image/jpeg"];
 
 let storage = new Storage();
 
 if (fullConfig.googleSvcAcctKey !== "") {
-  storage = new Storage({keyFilename: fullConfig.googleSvcAcctKey});
+  storage = new Storage({ keyFilename: fullConfig.googleSvcAcctKey });
 }
 
-const storageBucketName = fullConfig.storageBucketName
+const storageBucketName = fullConfig.storageBucketName;
 
-class UploadController {
-  async upload(req: express.Request, res: express.Response) {
-    // check first before proceeding
-    if (req.body.originalFilename === undefined || req.body.originalFilename === "") {
-      const result = {"status": "failure", "message": "Invalid original filename"};
-      return res.end(JSON.stringify(result));
-    }
+const UploadRequestQuerySchema = z.object({
+  originalFileName: z.string(),
+  base64File: zodBase64FileSchema(MAX_FILE_SIZE_MB),
+});
 
-    const MAX_UPLOAD_MB = 10;
-    if (req.body.base64File.length / 4 * 3 > MAX_UPLOAD_MB * 1024 * 1024) {
-      const result = {"status": "failure", "message": "File size too large"};
-      return res.end(JSON.stringify(result));
-    }
+const UploadImageRequestQuerySchema = z.object({
+  originalFileName: z.string(),
+  base64File: zodBase64FileSchema(MAX_IMAGE_SIZE_MB),
+});
+
+@JsonController("/api/upload")
+export class UploadController {
+  @Post()
+  @UseBefore(IsLoggedInMiddleware)
+  async upload(
+    @Req() req: express.Request,
+  ): Promise<Upload.uploadApiDomain.response> {
+    const input = UploadRequestQuerySchema.parse(req.body);
+
+    const {
+      buffer,
+      mimeType,
+      extension,
+    } = await this.validateAndExtractFileInfo(
+      input,
+      ALLOWED_FILE_MIME_TYPES,
+    );
+
+    const newFileName = makeRandomName() + "." + extension;
```

Comment:
would be nice to include original filename appended onto the random name. Not necessary by nice to have.